### PR TITLE
More CFn features tests

### PR DESCRIPTION
**Pull request by pinzon**
- [x] Drift detection test
- [x] "Linting error instead of trying to deploy" test
- [x] list-imports tests 
- [x] Notifications
- [x] Update and delete nested stacks
- [x] DeletePolicy: Retain
- [x] Stack policy to prevent deletion

### CODE REVIEW

File: tests/integration/cloudformation/api/test_stacks.py [hunk 1]
```diff
@@ -293,3 +294,160 @@ def stack_process_is_finished(cfn_client, stack_name):
         "PROGRESS"
         not in cfn_client.describe_stacks(StackName=stack_name)["Stacks"][0]["StackStatus"]
     )
+
+
+@pytest.mark.skip(reason="Not implemented")
+@pytest.mark.aws_verified
+def test_drift_detection_on_lambda(deploy_cfn_template, cfn_client, lambda_client, snapshot):
```
Comment (dominikschubert) [inline tests/integration/cloudformation/api/test_stacks.py:301]: just a mental note: Looking over this made me realize we probably should include this in our "resource testing framework" (i.e. include a test for this for each resource) when we start on the new providers.

File: tests/integration/cloudformation/api/test_stacks.py [hunk 2]
```diff
@@ -293,3 +294,160 @@ def stack_process_is_finished(cfn_client, stack_name):
...
+    stack_name = f"stack-{short_uid()}"
+    bad_template = {"Resources": "", "Outputs": ""}
+
+    with pytest.raises(botocore.exceptions.ClientError) as ex:
+        cfn_client.create_stack(StackName=stack_name, TemplateBody=json.dumps(bad_template))
+
+    error_response = ex.value.response
+
+    assert error_response["Error"]["Code"] == "ValidationError"
+    assert "Template format error" in error_response["Error"]["Message"]
```
Comment (dominikschubert) [inline tests/integration/cloudformation/api/test_stacks.py:333]: I'd encourage the use of a snapshot here. Might seem a bit unnecessary but asserts like this have failed us in the past unfortunately: `assert "Template format error" in error_response["Error"]["Message"]`

File: tests/integration/templates/lambda_url.yaml [hunk 3]
```diff
@@ -28,6 +29,10 @@ Resources:
           - Arn
       Handler: index.handler
       Runtime: python3.9
+      Environment:
+        Variables:
+          ENDPOINT_URL: aws.amazon.com
+
```
Comment (dominikschubert) [inline tests/integration/templates/lambda_url.yaml:35]: I think we should do this via a different template file. It doesn't really make sense for the other test case(s) using the template.

File: tests/integration/cloudformation/api/test_stacks.py [hunk 4]
```diff
@@ -293,3 +294,160 @@ def stack_process_is_finished(cfn_client, stack_name):
...
+        # Assert notifications of resources deleted
+        assert [message for message in messages.values() if "UPDATE_" in message["Message"]]
+
+        # Assert notifications of resources deleted
+        assert [message for message in messages.values() if "DELETE_" in message["Message"]]
+
+    try:
+        retry(_assert_messages, retries=10, sleep=2)
+    finally:
+        cfn_client.delete_stack(StackName=stack_name)
```
Comment (dominikschubert) [inline tests/integration/cloudformation/api/test_stacks.py:397]: We have a `cleanups` fixture that is better suited for this purpose.

It's just a list that you can fill with callbacks to be executed in reverse order after the test finishes.

The general pattern looks like this:

```python
def test_something(cleanups):
    result = client.create_something(...)
    cleanups.append(lambda: client.delete_something(Id=result['id'])

    ... rest of test
```

File: tests/integration/cloudformation/api/test_nested_stacks.py [hunk 5]
```diff
@@ -106,3 +107,57 @@ def test_nested_with_nested_stack(cfn_client, deploy_cfn_template, s3_client, s3
...
+        template_mapping={
+            "s3_bucket_url": f"/{bucket_name}/{key}",
+            "nested_bucket_name": altered_nested_bucket_name,
+        },
+        max_wait=120 if is_aws_cloud() else None,
+    )
+
+    assert s3_client.head_bucket(Bucket=altered_nested_bucket_name)
+
+    cfn_client.delete_stack(StackName=stack.stack_name)
```
Comment (dominikschubert) [inline tests/integration/cloudformation/api/test_nested_stacks.py:154]: FYI: there's a convenience function that you *can* use if you want

```python
stack = deploy_cfn_template(
...
)
stack.destroy()
```

**Review by dominikschubert — CHANGES_REQUESTED** [review]
Some minor suggestions regarding cleanup and templates.

File: tests/integration/cloudformation/api/test_stacks.py (hunk 1 above)
Comment (pinzon) [reply tests/integration/cloudformation/api/test_stacks.py]: I moved this test into its own file so we remember that it should be populated.

**Review by dominikschubert — APPROVED** [review]
LGTM :rocket:

**Pull request merged**